Auto register services

Nyholm opened the pull request:
This will auto register factories based on what services you have enabled.

mnapoli commented:
![Capture d’écran 2020-03-19 à 16 24 15](https://user-images.githubusercontent.com/720328/77083626-1ac5c600-69fe-11ea-8036-a3ac77bb50e9.png)

AFAICT this is the main change for users? It looks a bit more complex, what would we gain from that change? I'm not sure I understand the change TBH (the title mentions auto-registration, yet we still need to declare them in the config)

Nyholm commented:
> AFAICT this is the main change for users?

No, that is if you explicitly want to configure the SQS client. 

I added a new paragraph at the bottom of the readme. The readme is also updated. 
PR is ready for a review.

Nyholm commented on README.md:
```diff
@@ -160,9 +157,6 @@ framework:
-    sns: true # Register the SNS transport
-
 services:
     Aws\Sns\SnsClient:
```
There must be a service called `Aws\Sns\SnsClient` when you are using a DSN `sns://...`. With this PR you may now configure that with `bref_messenger.transports.sns.client: 'foobar_service'`

mnapoli replied:
In which cases can this be helpful?

Nyholm replied:
When you have multiple SNS clients or if you have existing services that you don't want to rename.

Nyholm commented on tests/Functional/Service/Sns/SnsTransportTest.php:
```diff
@@ -23,7 +23,7 @@ protected function getDefaultConfig(): array
     {
         /** @var SnsTransportFactory $factory */
-        $factory = $this->container->get(SnsTransportFactory::class);
+        $factory = $this->container->get('bref_messenger.transport.sns');
```
We now support multiple services to be defined with `SnsTransportFactory `

Nyholm commented on src/DependencyInjection/TransportProvider.php:
```diff
@@ -0,0 +1,41 @@
+            ],
+            'sqs' => [
+                'default_client' => SqsClient::class,
+                'package' => 'aws/aws-sdk-php',
```
Oh, this class looks a bit strange... It is like it is preparing to allow other packages than `aws/aws-sdk-php` to to be used with this bundle. 
=)

mnapoli commented on composer.json:
```diff
@@ -26,6 +25,7 @@
         "bref/bref": "^0.5.14"
     },
     "require-dev": {
+        "aws/aws-sdk-php": "^3.127",
```
Why move it to require-dev?

Nyholm replied:
I want to make a follow up PR to introduce Async AWS. See https://github.com/brefphp/symfony-messenger/pull/16#discussion_r395148988

With this factories and the TransportProvider we have detached ourselves from `aws/aws-sdk-php`. 

Yes, it is a BC break. But Im okey with this since nothing worked "out of the box" before, you still had to register a `Aws\Sqs\SqsClient` service.

mnapoli replied:
OK I see. I'm on the fence because this brings more complexity, and this mixes with a PR that isn't directly related.

What if this package always used async-aws? Would that work? From what I can picture in my head, this package can have its own SQS/SNS client and that wouldn't leak out to impact the user 🤔 

I'd rather avoid a situation where this package is compatible with both SDKs, because that means more maintenance effort and more effort for users as well.

mnapoli commented on src/Resources/config/services.yaml:
```diff
@@ -2,4 +2,4 @@ services:
     Bref\Symfony\Messenger\Service\SimpleBusDriver:
         arguments:
-            - '@logger'
+            - '@?logger'
```
Why do we make it optional?

Nyholm replied:
This service might not exist. That is why I made this optional.

mnapoli replied:
Oh ok I thought that service always existed in Symfony.

Nyholm commented:
Excellent questions!

> Why do we need to be able to disable the registration of transports?

If I have a transport with DSN `sqs://` and `sns://`, I may want to use the Symfony Messenger transport for SQS and the BrefMessenger transport for SNS.

If I haven't registered a `Aws\Sqs\SqsClient` service, then I will get a build error. If I have registered that service, then both Symfony Messenger transport and the BrefMessenger transport will be used for SQS. 

(I think it is both, it might be the one registered first)

To avoid this unwanted behavior, there is a way to opt-out of the auto registration. Both "globally" and for a specific type. 

> Why can we create a custom transport factory?

If I have two SqsClients with different config (ie two AWS accounts), I need a way to register them both.

Nyholm commented:
Is there anything else I can to to help to review this PR?

Nyholm commented:
Do you mind if I move forward with this?
I want to add support for async-aws

mnapoli commented:
Sorry for the delay.

> If I have a transport with DSN sqs:// and sns://, I may want to use the Symfony Messenger transport for SQS and the BrefMessenger transport for SNS.

OK I see. I really wish Symfony worked in less convoluted ways sometimes… Something where you explicitly define which transport to use instead of having something guess based on the DSN shape.

Would it make sense to add this to the documentation? (and document the use cases, not just the technical solutions)

> > Why can we create a custom transport factory?

> If I have two SqsClients with different config (ie two AWS accounts), I need a way to register them both.

That sounds like a very specific use case, I'm not sure we should aim to support it right now. People with this use case can probably setup things by themselves?

Can we find a way to open the code to those advanced use cases, instead of implementing (and supporting) them?

Nyholm commented:
The current implementation basically works as you expect. You install the bundle and one AWS SDK  and you are good to go. Everything works as expected. 

This bundle also handle all weird edge cases where you have multiple SDK's installed or the Symfony 5.1 SQS transport. And it handle those edge cases so it always works and is predictable. 

> Can we find a way to open the code to those advanced use cases, instead of implementing (and supporting) them?

I believe that this is exactly what I did.

Nyholm commented:
Im closing this as #22 was merged.
